app/grandchallenge/algorithms/forms.py

```diff
@@ -365,21 +366,87 @@ class Meta:
+        super().__init__(*args, user=user, **kwargs)
+        self._user = user
 
         self.fields["contact_email"].required = True
         self.fields["display_editors"].required = True
 
-        self.fields["job_requires_gpu_type"].choices = [
-            (choice.value, choice.label)
-            for choice in [GPUTypeChoices.NO_GPU, GPUTypeChoices.T4]
-        ]
+        self.fields["job_requires_gpu_type"].choices = (
+            self.selectable_gpu_type_choices
+        )
         self.fields["job_requires_memory_gb"].validators = [
             MinValueValidator(settings.ALGORITHMS_MIN_MEMORY_GB),
-            MaxValueValidator(settings.ALGORITHMS_MAX_MEMORY_GB),
+            MaxValueValidator(self.maximum_settable_memory_gb),
         ]
 
+    @property
+    def relevant_phases(self):
+        qs = get_objects_for_user(
+            self._user, "evaluation.create_phase_submission"
+        )
+        inputs = self.instance.inputs.all()
+        outputs = self.instance.outputs.all()
+        return qs.annotate(
+            total_algorithm_input_count=Count(
+                "algorithm_inputs", distinct=True
+            ),
+            total_algorithm_output_count=Count(
+                "algorithm_outputs", distinct=True
+            ),
+            relevant_algorithm_input_count=Count(
+                "algorithm_inputs",
+                filter=Q(algorithm_inputs__in=inputs),
+                distinct=True,
+            ),
+            relevant_algorithm_output_count=Count(
+                "algorithm_outputs",
+                filter=Q(algorithm_outputs__in=outputs),
+                distinct=True,
+            ),
+        ).filter(
+            submission_kind=SubmissionKindChoices.ALGORITHM,
+            public=True,
+            total_algorithm_input_count=len(inputs),
+            total_algorithm_output_count=len(outputs),
+            relevant_algorithm_input_count=len(inputs),
+            relevant_algorithm_output_count=len(outputs),
+        )
+
+    @property
+    def selectable_gpu_type_choices(self):
+        choices = [GPUTypeChoices.NO_GPU, GPUTypeChoices.T4]
+        choices.extend(
+            chain.from_iterable(
+                self.relevant_phases.aggregate(
+                    choices=ArrayAgg(
+                        "algorithm_selectable_gpu_type_choices", distinct=True
+                    )
+                )["choices"]
+            )
+        )
+        choices_set = set(choices)
+        return [
+            (choice.value, choice.label)
+            for choice in GPUTypeChoices
+            if choice in choices_set
+        ]
+
+    @property
+    def maximum_settable_memory_gb(self):
+        value = settings.ALGORITHMS_MAX_MEMORY_GB
+        maximum_in_phases = (
+            self.relevant_phases.order_by(
+                "-algorithm_maximum_settable_memory_gb"
+            )
+            .values_list("algorithm_maximum_settable_memory_gb", flat=True)
+            .first()
+        )
```

Comment:
This will still cause the new queries to go to the database twice. Once for the max memory, once for the selectable gpus. Did you consider doing both in a single aggregation, for instance using the `Max` method? See https://docs.djangoproject.com/en/5.1/topics/db/aggregation/#generating-aggregates-over-a-queryset

If that works, you could move the aggregation to `relevant_phases` (probably change the name as it would then return a map with two keys, max memory and selectable gpus), make that a `cached_property`, then have the two properties extract and clean the values returned from the database.

If that is unclear I can meet at 9am EU time on Thursday to explain what I mean, let me know!

Reply:
Cool use of `ArrayAgg` and `distinct`, didn't know about that either.

Reply:
I think I understand, and I tried that. But what I ended up with was something like this and when I timed it, it was slower:
```python
query = list(Phase.objects.annotate(
    max_memory=Max('algorithm_maximum_settable_memory_gb'),
    choices=ArrayAgg("algorithm_selectable_gpu_type_choices", distinct=True)
    ).values_list('max_memory', 'choices'))

max_value = query[0][0]  # The maximum value is constant across rows
choices = [row[1][0] for row in query]  # Extract array_field values

set(chain.from_iterable(choices))
```

Reply:
Hold on, I think I get my issue. I should not use annotate here, but aggregate instead.

Reply:
I think I got it now. Could you look at it again?

Reply:
Yes that is what I meant!